python-threatexchange/threatexchange/TE.py

```diff
@@ -13,15 +13,77 @@
 # General Python dependencies
 import os
 import re
-import urllib
-import urllib.error
+import requests
+
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
 import urllib.parse
-import urllib.request
+
+
+DEFAULT_TIMEOUT = 5  # seconds
+
+
+class TimeoutHTTPAdapter(HTTPAdapter):
+    """
+    Plug into requests to get a well-behaved session that does not wait for eternity.
+    H/T: https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/#setting-default-timeouts
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.timeout = DEFAULT_TIMEOUT
+        if "timeout" in kwargs:
+            self.timeout = kwargs["timeout"]
+            del kwargs["timeout"]
+        super().__init__(*args, **kwargs)
+
+    def send(self, request, **kwargs):
+        timeout = kwargs.get("timeout")
+        if timeout is None:
+            kwargs["timeout"] = self.timeout
+        return super().send(request, **kwargs)
+
+
+DEFAULT_TE_BASE_URL = "https://graph.facebook.com/v6.0"
+retry_strategy = Retry(
+    total=4,
+    status_forcelist=[429, 500, 502, 503, 504],
+    method_whitelist=["HEAD", "GET", "OPTIONS"],
+)
+adapter = TimeoutHTTPAdapter(timeout=60, max_retries=retry_strategy)
+
+
+def get_fb_graph_api():
+    """
+    Custom requests session that provides
+    - retries: 4 retries on GET requests for 429 or 5XX error codes
+    - timeout: 60 seconds. Can be adjusted at the fb_graph_api.get(.., timeout=FOO) level
+
+    Ideally, should be used within a context manager:
+    ```
+    with get_fb_graph_api() as fb_graph_api:
+        fb_graph_api.get()...
+    ```
+
+    If using without a context manager, ensure you end up calling close() on the returned value.
+
+    TODO: Identify if requests Session object can be used across threads in a
+    `concurrent.futures.ThreadPoolExecutor`
```

Comment:
I think I know why you've not faced issues.

tl;dr: the concern is when **multiple** hosts are connected to via the same PoolManager. We only connect to a single host.
---
PoolManager [1] transparently manages creating and destroying ConnectionPools. One ConnectionPool per host [2]. 

The issue reported on [github](https://github.com/urllib3/urllib3/issues/1252) talks about using multiple hosts and how the ConnectionPool retrieved from the PoolManager might get closed if too many hosts are connected to, frequently. Not an issue for us because we will not be connecting to multiple hosts.

Now the one on [stackoverflow](https://stackoverflow.com/questions/18188044/is-the-session-object-from-pythons-requests-library-thread-safe/32503750) is different. That is about the individual threads having access to Session internals and overwriting something like cookies, which it does without taking locks.

So, while the usage of requests is not thread-safe, the way we use it, it is. However, I'm still hesitant to re-use sessions across threads because of how our usage might evolve. I'd want to see if there are legit performance issues before we attempt re-using the same session across threads.

[1]: https://urllib3.readthedocs.io/en/latest/reference/urllib3.poolmanager.html
[2]: https://urllib3.readthedocs.io/en/latest/reference/urllib3.connectionpool.html